src/analysis/special_functions/arsinh.lean

```diff
@@ -0,0 +1,166 @@
+/-
+Copyright (c) 2020 James Arthur. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: James Arthur, Chris Hughes, Shing Tak Lam.
+-/
+import analysis.special_functions.trigonometric
+import data.real.basic
+import analysis.special_functions.pow
+
+open real
+local attribute [pp_nodot] real.log
+noncomputable theory
+
+/-!
+# Inverse of the sinh function
+
+In this file we prove that sinh is bijective and hence has an
+inverse, arsinh.
+
+## Main Results
+
+- `sinm_injective`: The proof that `sinm` is injective
+- `sinm_surjective`: The proof that `sinm` is surjective
+- `sinm_bijective`: The proof `sinm` is bijective
+- `arsinh`: The inverse function of `sinm`
+
+## Tags
+
+arsinh, sinh injective, sinh bijective, sinh surjective
+-/
+
+
+/-- `arsinh` is defined using a logarithm-/
+def arsinh (x : ℝ) := log (x + sqrt(1 + x^2))
+
+/-- `sinh` is injective, `∀ a b, sinh a = sinh b → a = b` -/
+lemma sinh_injective : function.injective sinh := sinh_strict_mono.injective
+
+private lemma aux_lemma (x : ℝ) : 1/(x + (1 + x^2).sqrt) = - x + (1 + x^2).sqrt :=
+begin
+  have H : (x - (1 + x^2).sqrt)/((x - (1 + x^2).sqrt) * (x + (1 + x^2).sqrt)) = -x + (1 + x^2).sqrt,
+  { have G : (x - (1 + x^2).sqrt) * (x + (1 + x^2).sqrt) = -1,
+    { ring,
+      rw sqr_sqrt,
+      { ring },
+      { linarith [pow_two_nonneg x]} },
+    rw G,
+    ring },
+  rw [division_def, mul_inv', ←mul_assoc, mul_comm (x - sqrt (1 + x ^ 2)), inv_mul_cancel] at H,
+  rw one_mul at H,
+  rw [division_def, one_mul],
+  exact H,
+  { intros fx,
+    rw sub_eq_zero at fx,
+    have fx_sq : x^2 = (sqrt (1 + x ^ 2)) ^ 2 := by { rw fx.symm },
+    rw sqr_sqrt at fx_sq,
+    linarith,
+    have G : 0 ≤ x^2 := by {apply pow_two_nonneg},
+    linarith }
+end
+
+private lemma b_lt_sqrt_b_sq_add_one (b : ℝ) : b < sqrt(b^2 + 1) :=
+begin
+  by_cases hb : 0 ≤ b,
+  conv { to_lhs, rw ← sqrt_sqr hb },
+  rw sqrt_lt,
+  linarith,
```

Comment:
similar here. IIRC `linarith` closes the current goal and fails if it can't do that, which means you have multiple goals here.

Reply:
resolved